PR: https://github.com/odoo/odoo/pull/63192

From: 3d5e912da785bb6cb7409491041a2492cd358be3
From: Kevin Baptiste
Date: 2021-02-05 13:00:09

Breaking data model changes scores: del:3 + add:0, change matches:
-    work_location = fields.Char('Work Location')
-    work_location = fields.Char(readonly=True)
-    work_location = fields.Char(related='employee_id.work_location', readonly=False, related_sudo=False)

Total Changes: 30

[IMP] hr: add a model for work locations

The work location of an employee was previously a fields.Char. That was
changed to a Many2one field to a the new model called work.location.
That had some consequences in other modules that had to be slightly
adjusted.  The modules impacted were hr, hr_appraisal, hr_payroll.

Task id: 2335646

closes odoo/odoo#63192

Related: odoo/enterprise#15235
Related: odoo/upgrade#2020
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr/models/__init__.py
+++ b/addons/hr/models/__init__.py
@@ -8,6 +8,7 @@ from . import hr_employee_public
 from . import hr_department
 from . import hr_job
 from . import hr_plan
+from . import hr_work_location
 from . import mail_alias
 from . import mail_channel
 from . import models

--- a/addons/hr/models/hr_employee_base.py
+++ b/addons/hr/models/hr_employee_base.py
@@ -27,7 +27,8 @@ class HrEmployeeBase(models.AbstractModel):
     work_phone = fields.Char('Work Phone', compute="_compute_phones", store=True, readonly=False)
     mobile_phone = fields.Char('Work Mobile')
     work_email = fields.Char('Work Email')
-    work_location = fields.Char('Work Location')
+    work_location_id = fields.Many2one('hr.work.location', 'Work Location', compute="_compute_work_location_id", store=True, readonly=False,
+    domain="[('address_id', '=', address_id), '|', ('company_id', '=', False), ('company_id', '=', company_id)]")
     user_id = fields.Many2one('res.users')
     resource_id = fields.Many2one('resource.resource')
     resource_calendar_id = fields.Many2one('resource.calendar', domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]")
@@ -162,6 +163,11 @@ class HrEmployeeBase(models.AbstractModel):
                     icon = 'presence_undetermined'
             employee.hr_icon_display = icon
 
+    @api.depends('address_id')
+    def _compute_work_location_id(self):
+        to_reset = self.filtered(lambda e: e.address_id != e.work_location_id.address_id)
+        to_reset.work_location_id = False
+
     @api.model
     def _get_employee_working_now(self):
         working_now = []

--- a/addons/hr/models/hr_employee_public.py
+++ b/addons/hr/models/hr_employee_public.py
@@ -24,7 +24,7 @@ class HrEmployeePublic(models.Model):
     mobile_phone = fields.Char(readonly=True)
     work_phone = fields.Char(readonly=True)
     work_email = fields.Char(readonly=True)
-    work_location = fields.Char(readonly=True)
+    work_location_id = fields.Many2one(readonly=True)
     user_id = fields.Many2one(readonly=True)
     resource_id = fields.Many2one(readonly=True)
     resource_calendar_id = fields.Many2one(readonly=True)

--- a/None
+++ b/addons/hr/models/hr_work_location.py
@@ -0,0 +1,15 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class WorkLocation(models.Model):
+    _name = "hr.work.location"
+    _description = "Work Location"
+    _order = 'name'
+
+    active = fields.Boolean(default=True)
+    name = fields.Char(string="Work Location", required=True)
+    company_id = fields.Many2one('res.company', required=True, default=lambda self: self.env.company)
+    address_id = fields.Many2one('res.partner', required=True, string="Work Address", domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]")

--- a/addons/hr/models/res_users.py
+++ b/addons/hr/models/res_users.py
@@ -21,7 +21,7 @@ class User(models.Model):
     category_ids = fields.Many2many(related='employee_id.category_ids', string="Employee Tags", readonly=False, related_sudo=False)
     department_id = fields.Many2one(related='employee_id.department_id', readonly=False, related_sudo=False)
     address_id = fields.Many2one(related='employee_id.address_id', readonly=False, related_sudo=False)
-    work_location = fields.Char(related='employee_id.work_location', readonly=False, related_sudo=False)
+    work_location_id = fields.Many2one(related='employee_id.work_location_id', readonly=False, related_sudo=False)
     employee_parent_id = fields.Many2one(related='employee_id.parent_id', readonly=False, related_sudo=False)
     coach_id = fields.Many2one(related='employee_id.coach_id', readonly=False, related_sudo=False)
     address_home_id = fields.Many2one(related='employee_id.address_home_id', readonly=False, related_sudo=False)
@@ -138,7 +138,7 @@ class User(models.Model):
             'visa_expire',
             'visa_no',
             'work_email',
-            'work_location',
+            'work_location_id',
             'work_phone',
             'certificate',
             'study_field',
